internal/client/client_interceptor.go

```diff
@@ -34,10 +37,84 @@ func NewInterceptedClient(client Client) Client {
 }
 
 func (r interceptedClient) SendRequest(ctx context.Context, addr string, req *tikvrpc.Request, timeout time.Duration) (*tikvrpc.Response, error) {
+	// Build the resource control interceptor.
+	resourceGroupName := req.GetResourceGroupName()
+	var finalInterceptor interceptor.RPCInterceptor = buildResourceControlInterceptor(ctx, req, resourceGroupName)
+	// Chain the interceptors if there are multiple interceptors.
 	if it := interceptor.GetRPCInterceptorFromCtx(ctx); it != nil {
-		return it(func(target string, req *tikvrpc.Request) (*tikvrpc.Response, error) {
+		if finalInterceptor != nil {
+			finalInterceptor = interceptor.ChainRPCInterceptors(finalInterceptor, it)
+		} else {
+			finalInterceptor = it
+		}
+	}
+	if finalInterceptor != nil {
+		return finalInterceptor(func(target string, req *tikvrpc.Request) (*tikvrpc.Response, error) {
 			return r.Client.SendRequest(ctx, target, req, timeout)
 		})(addr, req)
 	}
 	return r.Client.SendRequest(ctx, addr, req, timeout)
 }
+
+var (
+	resourceControlSwitch      atomic.Value
+	resourceControlInterceptor client.ResourceGroupKVInterceptor
+)
+
+// EnableResourceControl enables the resource control.
+func EnableResourceControl() {
+	resourceControlSwitch.Store(true)
+}
+
+// DisableResourceControl disables the resource control.
+func DisableResourceControl() {
+	resourceControlSwitch.Store(false)
+}
+
+// SetResourceControlInterceptor sets the interceptor for resource control.
+func SetResourceControlInterceptor(interceptor client.ResourceGroupKVInterceptor) {
```

Comment:
This function in the internal, cannot be called?

Reply:
Unless it's called inside the `client` package, otherwise I think it has to be exported.

Reply:
Yes, I call it in `tidb` package.